### PR TITLE
FEATURE: Add bulk action to bookmark

### PR DESCRIPTION
**Pull request by branquinhoaa**
This PR aims to add bulk actions to the user's bookmarks.

After this feature, all users should be able to select multiple bookmarks and perform the actions of "deleting" or "clear reminders" as shown below:

bookmarks

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/bookmark-list.hbs
Comment (branquinhoaa) [inline app/assets/javascripts/discourse/app/components/bookmark-list.hbs:10]: This follows the same pattern as the bulk topic selection.

File: app/assets/javascripts/discourse/app/components/bookmark-list.hbs
Comment (branquinhoaa) [inline app/assets/javascripts/discourse/app/components/bookmark-list.hbs:59]: Displays a checkbox for each bookmark so the user can choose.

File: app/assets/javascripts/discourse/app/controllers/user-activity-bookmarks.js
Comment (branquinhoaa) [inline app/assets/javascripts/discourse/app/controllers/user-activity-bookmarks.js:31]: I'm reusing the same helper used for topics.

File: app/assets/javascripts/discourse/app/models/bookmark.js
Comment (branquinhoaa) [inline app/assets/javascripts/discourse/app/models/bookmark.js:45]: Added a method to support bulk operations on bookmarks.

File: app/assets/javascripts/discourse/app/raw-templates/flat-button.hbr
Comment (branquinhoaa) [inline app/assets/javascripts/discourse/app/raw-templates/flat-button.hbr:1]: This was just a minor refactoring to get the title dynamically instead of always getting the hardcoded for topics.

File: app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js
Comment (branquinhoaa) [inline app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js:50]: This is the test I mentioned before; it is currently flakey because of the dialog service.

File: app/controllers/bookmarks_controller.rb
Comment (branquinhoaa) [inline app/controllers/bookmarks_controller.rb:78]: Add action to support bulk actions for bookmarks.

File: lib/bookmark_reminder_notification_handler.rb
Comment (branquinhoaa) [inline lib/bookmark_reminder_notification_handler.rb:27]: I'm using this method to clear the reminders for bookmarks.

File: spec/requests/bookmarks_controller_spec.rb
Comment (branquinhoaa) [inline spec/requests/bookmarks_controller_spec.rb:10]: I had to move this inside the describe blocks to add tests for signed out users.

**Review by branquinhoaa — COMMENTED** [review]
Self-review.

File: app/assets/javascripts/discourse/app/components/bookmark-list.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/bookmark-list.hbs:25]: Can you try to use the Glimmer component directly here? For context, raw handlebars components are legacy elements, we are rtying to move away from them. (There's no way for you to have known this before...)

I suspect this may also help with the dialog test issue.

File: app/assets/javascripts/discourse/app/components/bookmark-list.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/bookmark-list.hbs:38]: Same here, het's try to avoid any raw components.

File: app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js
Comment (pmusaraj) [reply app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js:50]: Yeah, this is tricky. I am not sure it ahs to do with the dialog service, that service should be available at all times. I suspect it has to do with raw handlebars and the fact that here two tests are loading the same endpoint. 

Once we move away from handlebars, this should clear. If it doesn't, we can sidestep tihs issue by having just one test that covers both the deleting and the reminder clearing checks.

File: app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js
Comment (pmusaraj) [inline app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js:23]: Very minor, but you likely don't ned to set moderator to true here. Regular users should be able to update their own bookmarks.

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (pmusaraj) [inline app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js:58]: For the "delete bookmark" use case, it may be better to use the delete confirm dialog.

File: spec/requests/bookmarks_controller_spec.rb
Comment (pmusaraj) [inline spec/requests/bookmarks_controller_spec.rb:171]: Instead of stubbing the guardian check, it might be better to run this call as the correct user. And also run it as a second user (which should fail to update the bookmarks for user 1.)

File: spec/requests/bookmarks_controller_spec.rb
Comment (pmusaraj) [inline spec/requests/bookmarks_controller_spec.rb:187]: Same here, given this is a destructire operation, we want to make sure the only right users can do this.

File: lib/bookmarks_bulk_action.rb
Comment (pmusaraj) [inline lib/bookmarks_bulk_action.rb:16]: I don't think we need to add an extension point at this time.

File: app/controllers/bookmarks_controller.rb
Comment (pmusaraj) [inline app/controllers/bookmarks_controller.rb:93]: I think we can keep the bookmark operation types separate from the topic bulk action operation types here.

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (pmusaraj) [inline spec/lib/bookmarks_bulk_action_spec.rb:12]: Here too we should also have a few tests with the bookmark edit permission.

**Review by pmusaraj — COMMENTED** [review]
Thanks @branquinhoaa, this is a great start. I left some comments from a first review pass. 

I also took it for a spin locally, and it works well. The main comment is regarding raw handlebars templates, we should try and avoid them as much as possible. (I know you're basing this off the existing raw templates, but that's legacy code that we plan to retire in the near future.)

File: app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js
Comment (branquinhoaa) [reply app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js]: I've updated the templates, but the test was still flakey, so I combined all the tests into one, as you suggested.

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (branquinhoaa) [reply spec/lib/bookmarks_bulk_action_spec.rb]: I checked the [bookmark guardian](https://github.com/discourse/discourse/blob/main/lib/guardian/bookmark_guardian.rb) implementation, and it seems that currently, all users who can edit can also delete. So, the tests will be the same. I still added them to avoid future regressions, though.

**Comment by branquinhoaa** [comment]
@pmusaraj Thanks for the review! 

I've applied your suggestions and left some comments on them. 

Could you take another look? Please let me know if you have anything else in mind.

File: app/assets/javascripts/discourse/app/components/bookmark-list.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/bookmark-list.hbs:38]: For these two buttons, we can use `<DButton>` and `@action` since this is all in a regular Ember component.

File: app/assets/javascripts/discourse/app/raw-views/bookmark-bulk-select-dropdown.gjs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/raw-views/bookmark-bulk-select-dropdown.gjs:6]: Does the PR use this raw component now?

File: app/assets/javascripts/discourse/app/raw-templates/flat-button.hbr
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/raw-templates/flat-button.hbr:1]: Probably don't need to include this change either now that bookmarks don't use the raw button.

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (pmusaraj) [inline app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js:39]: Can you try using the `content: ` property? `computeContent` is marked as deprecated, although, we seem to be silencing those deprecations. I am curious if actually switching to `content: ` will address the test issues as well.

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (pmusaraj) [inline app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js:102]: I'm not sure we need to toggle the bulk select state here. Maybe the user wants to continue bulk selecting and editing other bookmarks?

File: lib/bookmarks_bulk_action.rb
Comment (pmusaraj) [inline lib/bookmarks_bulk_action.rb:42]: This means that if you have one id of a bookmark that you can't edit, the operation will stop there and ignore the remaining bookmarks? We're not doing the same for deletes... it would be nice to have something consistent.

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (pmusaraj) [inline spec/lib/bookmarks_bulk_action_spec.rb:20]: This test case is a duplicate of the previous one.

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (pmusaraj) [inline spec/lib/bookmarks_bulk_action_spec.rb:10]:
```suggestion
    describe "when user is not the bookmark owner" do
```

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (pmusaraj) [inline spec/lib/bookmarks_bulk_action_spec.rb:40]: Duplicate as well.

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (pmusaraj) [inline spec/lib/bookmarks_bulk_action_spec.rb:51]:
```suggestion
  describe "#clear_reminder" do
```

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (pmusaraj) [inline spec/lib/bookmarks_bulk_action_spec.rb:54]:
```suggestion
    describe "when user is not the bookmark owner" do
```

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (pmusaraj) [inline spec/lib/bookmarks_bulk_action_spec.rb:62]:
```suggestion
    describe "when user is the bookmark owner" do
```

File: app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js
Comment (pmusaraj) [inline app/assets/javascripts/discourse/tests/acceptance/user-bookmark-bulk-actions-test.js:39]: I don't think checking the presence of the options is necessary, since we're selecting each option just below and therefore implicitly ensuring their presence.

**Review by pmusaraj — COMMENTED** [review]
Thanks for the updates, a few more comments.

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (branquinhoaa) [reply app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js:102]: It seems that the toggle is not necessary since the router refresh is already cleaning up the selection. 

In terms of user experience, however,  I can't see a specific case where keeping the selection would be helpful for the current "bookmark's bulk operations." 

If the user deletes the bookmark, he can't clean up the reminders, and if they spend effort cleaning the reminders, I don't think they will want to delete the same bookmarks in a subsequent operation. 

Are you thinking about a specific workflow? Maybe I'm missing something; I will be happy to fix this; just thought it could be something to think about and discuss :)

File: spec/lib/bookmarks_bulk_action_spec.rb
Comment (branquinhoaa) [reply spec/lib/bookmarks_bulk_action_spec.rb:20]: Yes, it seems right now the users who can delete are the same as the ones who can edit:
Screenshot 2024-05-06 at 22 13 10

I was unsure if adding tests for users who could edit would make sense since they would be the same. However, as you might change this in the future, I duplicated the tests. I'm removing the duplication, and the test can be added when things change.

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (branquinhoaa) [reply app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js:39]: Unfortunately, this update didn't help the test issue. I still see the error sometimes if I add multiple tests 🤔

File: lib/bookmarks_bulk_action.rb
Comment (branquinhoaa) [reply lib/bookmarks_bulk_action.rb:42]: When I was adding the tests, I noticed that the exception was being raised for delete action because of how [BookmarkManager](https://github.com/discourse/discourse/blob/2347ff707455e6da6447f12f3715a3bded408cc0/lib/bookmark_manager.rb#L130C1-L131C1) works. The BookmarkManager is raising the Invalid access error for delete operations. I wanted to do the same for clear reminders to keep it consistent. 
If you think it is better, I can remove the error, but I wanted to give you the full context of why I added it.

File: app/assets/javascripts/discourse/app/components/bookmark-list.hbs
Comment (branquinhoaa) [reply app/assets/javascripts/discourse/app/components/bookmark-list.hbs:38]: Great improvement, thank you 😄

File: app/assets/javascripts/discourse/app/raw-views/bookmark-bulk-select-dropdown.gjs
Comment (branquinhoaa) [reply app/assets/javascripts/discourse/app/raw-views/bookmark-bulk-select-dropdown.gjs:6]: no! good catch 👯

**Comment by branquinhoaa** [comment]
Hey @pmusaraj, thank you again for your time here; I really appreciate it.  

I added some notes for context - just to clarify some of the decisions I made, but please let me know if you disagree 😃

File: app/assets/javascripts/discourse/app/raw-templates/bookmark-bulk-select-dropdown.hbr
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/raw-templates/bookmark-bulk-select-dropdown.hbr:1]: This can be dropped as well?

File: lib/bookmarks_bulk_action.rb
Comment (pmusaraj) [reply lib/bookmarks_bulk_action.rb:42]: Ah, that makes sense, we are doing the same for deletes. I think you're right to keep the consistency. Maybe we move the check inside of `BookmarkReminderNotificationHandler.new(b).clear_reminder` so that individual reminder clearances also get the access check?

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (pmusaraj) [inline app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js:15]: I don't think we need this for bookmarks. In the topics list, this is linked to a plugin API endpoint that lets plugins add their own buttons and actions.

**Review by pmusaraj — APPROVED** [review]
A few minor cleanups, but otherwise this is looking great, thanks!

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (pmusaraj) [reply app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js]: Yeah, quiet strange. I think we can let this be for now.

File: app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js
Comment (pmusaraj) [reply app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js]: Yes, if you have say 20 bookmarks, you might want to: 

- select 5, clear reminders 
- select 2, delete
- continue with more selections

Comment (branquinhoaa) [reply app/assets/javascripts/select-kit/addon/components/bulk-select-bookmarks-dropdown.js]: Ooh, sorry for the confusion! I misunderstood your suggestion. I thought you wanted to keep the checkboxes selected, which is not what you are suggesting 😅 
So, this is what we have  at this point - I guess it matches your initial idea:
Bookmarks - Activity - amandaalves15 - Discourse 2024-05-07 at 10 57 17 PM

File: lib/bookmarks_bulk_action.rb
Comment (branquinhoaa) [reply lib/bookmarks_bulk_action.rb:42]: I started this update and noticed that it would affect places where I'm unsure of the impact.

To add this check inside ```BookmarkReminderNotificationHandler``` I need to pass the user as an argument when a new ```BookmarkReminderNotificationHandler``` is initialized. This would require updates in 3 files and updates in its correspondent tests.

I think it would be better to change this in a separate PR to keep the scope of this one smaller. It would also be easier to change and possibly revert if something goes wrong. 

Let me know your thoughts.

**Comment by branquinhoaa** [comment]
Hey @pmusaraj! These are done 😄 let me know what you think about the update for ```BookmarkReminderNotificationHandler```

File: lib/bookmarks_bulk_action.rb
Comment (pmusaraj) [reply lib/bookmarks_bulk_action.rb:42]: Apologies for the delay, Amanda! 

> I think it would be better to change this in a separate PR to keep the scope of this one smaller. It would also be easier to change and possibly revert if something goes wrong.
> 

Yes, this makes sense to me!

**Comment by pmusaraj** [comment]
@branquinhoaa I ran the specs here multiple times today (because I wanted to merge the PR), but they fail regularly, at the same place. 🤔

**Comment by branquinhoaa** [comment]
@pmusaraj I will check today!

**Comment by branquinhoaa** [comment]
Hey @pmusaraj, I have checked [my commit](https://github.com/discourse/discourse/pull/26856/commits/361352261cb68fb3c2a0e024f607a451a9658855) where it started failing, and I didn't find anything that could justify the error, so I'm updating my branch with the latest changes pushed to main to see if this could help. 

I have enabled Github actions on my fork to check if tests are passing. It is taking a while because of my GitHub plan. I will let you know the results once it is done 👍 

Do you know if there is an easier way to run these tests locally? I saw quite a bit of setup on the CI file.

**Comment by pmusaraj** [comment]
Sorry for my delay here @branquinhoaa, I think we're good to merge this now and keep an eye on flakey tests. 

Thanks for your work! LGTM! 

I'll let you click the Merge button, you should have that power now 😄

**Comment by branquinhoaa** [comment]
Yeey!! I just merged my first PR into Discourse's codebase 🥳  thank you @pmusaraj

**Pull request merged**
